src/app/modules/payments/stripe.utils.ts

```diff
@@ -0,0 +1,256 @@
+ * @returns err(ComputePaymentStateError) if there is an error in computing the state
+ */
+export const computePaymentState = (
+  events: Stripe.Event[],
+): Result<
+  Pick<Payment, 'status' | 'chargeIdLatest'>,
+  ComputePaymentStateError
+> => {
+  const logMeta = {
+    action: 'computePaymentState',
+    events,
+  }
+
+  // Helper to get the charge id from a nested charge object.
+  const getChargeId = (charge: string | Stripe.Charge): string =>
+    typeof charge === 'string' ? charge : charge.id
+
+  // We only care about charge and dispute events for computing payment status.
+  // The event types are:
+  // - charge.(failed|pending|refunded|succeeded)
+  // - charge.dispute.(closed|created|funds_reinstated|funds_withdrawn|updated)
+  const chargeEvents = (events as Stripe.DiscriminatedEvent[]).filter(
+    (
+      event: Stripe.DiscriminatedEvent,
+    ): event is
+      | Stripe.DiscriminatedEvent.ChargeEvent
+      | Stripe.DiscriminatedEvent.ChargeDisputeEvent =>
+      event.type === 'charge.failed' ||
+      event.type === 'charge.pending' ||
+      event.type === 'charge.refunded' ||
+      event.type === 'charge.succeeded' ||
+      event.type.startsWith('charge.dispute.'),
+  )
+
+  // Step 1: State machine for computing the current value of the payment state.
+  // A "null" status represents an unknown state, and should be recovered from later.
+  let status: Payment['status'] | null = PaymentStatus.Pending
+  let chargeIdLatest: Payment['chargeIdLatest']
+
+  for (const event of chargeEvents) {
+    switch (status) {
+      case PaymentStatus.Pending:
+      case PaymentStatus.Failed:
+        if (event.type === 'charge.failed') {
+          status = PaymentStatus.Failed
+          chargeIdLatest = event.data.object.id
+        } else if (event.type === 'charge.succeeded') {
+          status = PaymentStatus.Succeeded
+          chargeIdLatest = event.data.object.id
+        } else {
+          status = null
+        }
+        break
+      // Once the payment is successful, ensure that all future events are
+      // related to the latest charge. Otherwise, the status should be in an
+      // unknown state.
+      case PaymentStatus.Succeeded:
+      case PaymentStatus.PartiallyRefunded:
+        if (
+          event.type === 'charge.refunded' &&
+          event.data.object.id === chargeIdLatest
+        ) {
+          status =
+            event.data.object.amount_captured ===
+            event.data.object.amount_refunded
+              ? PaymentStatus.FullyRefunded
+              : PaymentStatus.PartiallyRefunded
+        } else if (
+          event.type === 'charge.dispute.created' &&
+          getChargeId(event.data.object.charge) === chargeIdLatest
+        ) {
+          status = PaymentStatus.Disputed
+        } else {
+          status = null
+        }
+        break
+      case PaymentStatus.Disputed:
+        if (
+          (event.type === 'charge.dispute.funds_withdrawn' ||
```

Comment:
this one will have a tradeoff since it's not a custom typeguard. We rely on the equality check to ensure that `event.data.object` is a `Stripe.Dispute` which makes the `.charge` attribute available later on:

```typescript
getChargeIdFromNestedCharge(event.data.object.charge) === chargeIdLatest
```

However it's possible to use `.includes` and then do something like
```typescript
getChargeIdFromNestedCharge((event.data.object as Stripe.Dispute).charge) === chargeIdLatest
```
what do you think is better?

Reply:
TIL, type narrowing isn't supported on [`Array.prototype.includes`](https://github.com/microsoft/TypeScript/issues/15048). 

I think for this specific case, we do make use of type narrowing, and thus it actually changes the intended typecheck behaviour at the cost of being terse. That isn't what we want.

With that, I think we should keep the original implementation. 🙈